nutils/util.py

```diff
@@ -437,6 +438,44 @@ def binaryfile(path):
     raise TypeError('binaryfile requires a path-like or file-like argument')
 
 
+def defaults_from_env(f):
+    '''Decorator for changing function defaults based on environment.
+
+    This decorator searches the environment for variables matching the pattern
+    ``NUTILS_MYPARAM``, where ``myparam`` is a parameter of the decorated
+    function. Only parameters with type annotation and a default value are
+    considered, and the string value is deserialized using stringly. In case
+    deserialization fails, a warning is emitted and the original default is
+    maintained.
+    '''
+
+    sig = inspect.signature(f)
+    params = []
+    changed = False
+    for param in sig.parameters.values():
+        envname = f'NUTILS_{param.name.upper()}'
+        if envname in os.environ and param.annotation != param.empty and param.default != param.empty:
+            try:
+                v = stringly.loads(param.annotation, os.environ[envname])
+            except Exception as e:
+                warnings.warn(f'ignoring environment variable {envname}: {e}')
+            else:
+                param = param.replace(default=v)
+                changed = True
+        params.append(param)
+    if not changed:
+        return f
+    sig = sig.replace(parameters=params)
+    def defaults_from_env(*args, **kwargs):
+        bound = sig.bind(*args, **kwargs)
+        bound.apply_defaults()
+        return f(*bound.args, **bound.kwargs)
+    defaults_from_env.__signature__ = sig
+    defaults_from_env.__wrapped__ = f
+    defaults_from_env.__doc__ = f.__doc__
```

Comment:
Why not `@functools.wraps(f)` decorate the wrapper? `@functools.wraps` copies more than `__wrapped__` and `__doc__`, for example: `__module__`, `__name__`, `__qualname__`.

Reply:
It does not introduce the new signature though. Better to combine wraps with assigning `__signature__` only?

Reply:
That's better IMO. If I'm not mistaken, Sphinx will ignore functions in a module if the `__module__` attribute doesn't match the global `__name__`.

Reply:
fixed